File: src/main/java/toby/command/commands/moderation/AdjustUserCommand.java
package toby.command.commands.moderation;

import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.Mentions;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.interactions.commands.OptionMapping;
import net.dv8tion.jda.api.interactions.commands.OptionType;
import net.dv8tion.jda.api.interactions.commands.build.OptionData;
import toby.command.CommandContext;
import toby.jpa.dto.UserDto;
import toby.jpa.service.IUserService;

import java.util.List;
import java.util.Optional;

import static toby.command.ICommand.invokeDeleteOnMessageResponse;
import static toby.helpers.UserDtoHelper.userAdjustmentValidation;
import static toby.jpa.dto.UserDto.Permissions.*;

public class AdjustUserCommand implements IModerationCommand {

    public static final String MODIFIER = "modifier";
    private final IUserService userService;
    private final String PERMISSION_NAME = "name";
    private final String USERS = "users";

    public AdjustUserCommand(IUserService userService) {
        this.userService = userService;
    }

    @Override
    public void handle(CommandContext ctx, UserDto requestingUserDto, Integer deleteDelay) {
        SlashCommandInteractionEvent event = ctx.getEvent();
        event.deferReply().queue();
        final Member member = ctx.getMember();
        Guild guild = event.getGuild();

        List<Member> mentionedMembers = channelAndArgumentValidation(event, requestingUserDto, guild.getOwner(), member, deleteDelay);
        if (mentionedMembers == null) return;

        mentionedMembers.forEach(targetMember -> {
            UserDto targetUserDto = userService.getUserById(targetMember.getIdLong(), targetMember.getGuild().getIdLong());
            //Check to see if the database contained an entry for the user we have made a request against
            if (targetUserDto != null) {
                boolean isSameGuild = requestingUserDto.getGuildId().equals(targetUserDto.getGuildId());
                boolean requesterCanAdjustPermissions = userAdjustmentValidation(requestingUserDto, targetUserDto) || member.isOwner();
                if (requesterCanAdjustPermissions && isSameGuild) {
                    validateArgumentsAndUpdateUser(event, targetUserDto, member.isOwner(), deleteDelay);
                    event.getHook().sendMessageFormat("Updated user %s's permissions", targetMember.getEffectiveName()).queue(invokeDeleteOnMessageResponse(deleteDelay));
                } else
                    event.getHook().sendMessageFormat("User '%s' is not allowed to adjust the permissions of user '%s'.", member.getEffectiveName(), targetMember.getEffectiveName()).queue(invokeDeleteOnMessageResponse(deleteDelay));

            } else {
                createNewUser(event, targetMember, deleteDelay);
            }
        });


    }

    private void validateArgumentsAndUpdateUser(SlashCommandInteractionEvent event, UserDto targetUserDto, Boolean isOwner, int deleteDelay) {
        Optional<String> permissionOptional = Optional.ofNullable(event.getOption(PERMISSION_NAME)).map(OptionMapping::getAsString);

        if (permissionOptional.isEmpty()) {
            event.getHook().sendMessage("You did not mention a valid permission to update").setEphemeral(true).queue(invokeDeleteOnMessageResponse(deleteDelay));
            return;
        }
        Optional<Integer> modifierOptional = Optional.ofNullable(event.getOption(MODIFIER)).map(OptionMapping::getAsInt);
        modifierOptional.ifPresent(targetUserDto::setInitiativeModifier);
        switch (UserDto.Permissions.valueOf(permissionOptional.get().toUpperCase())) {
            case MUSIC -> targetUserDto.setMusicPermission(!targetUserDto.hasMusicPermission());
            case DIG -> targetUserDto.setDigPermission(!targetUserDto.hasDigPermission());
            case MEME -> targetUserDto.setMemePermission(!targetUserDto.hasMemePermission());
            case SUPERUSER -> {
                if (isOwner) targetUserDto.setSuperUser(!targetUserDto.isSuperUser());
            }
        }

        userService.updateUser(targetUserDto);
    }

    private void createNewUser(SlashCommandInteractionEvent event, Member targetMember, int deleteDelay) {
        //Database did not contain an entry for the user we have made a request against, so make one.
        UserDto newDto = new UserDto();
        newDto.setDiscordId(targetMember.getIdLong());
        newDto.setGuildId(targetMember.getGuild().getIdLong());
        userService.createNewUser(newDto);
        event.getHook().sendMessageFormat("User %s's permissions did not exist in this server's database, they have now been created", targetMember.getEffectiveName()).queue(invokeDeleteOnMessageResponse(deleteDelay));
    }

    private List<Member> channelAndArgumentValidation(SlashCommandInteractionEvent event, UserDto requestingUserDto, Member member, Member guildOwner, int deleteDelay) {
        if (!member.isOwner() && !requestingUserDto.isSuperUser()) {
            event.getHook().sendMessage(getErrorMessage(guildOwner.getEffectiveName())).setEphemeral(true).queue(invokeDeleteOnMessageResponse(deleteDelay));
            return null;
        }

        Optional<List<Member>> mentionedMembersOptional = Optional.ofNullable(event.getOption(USERS)).map(OptionMapping::getMentions).map(Mentions::getMembers);
        if (mentionedMembersOptional.isEmpty()) {
            event.getHook().sendMessage("You must mention 1 or more Users to adjust permissions of").setEphemeral(true).queue(invokeDeleteOnMessageResponse(deleteDelay));
            return null;
        }

        if (Optional.ofNullable(event.getOption(PERMISSION_NAME)).map(OptionMapping::getAsString).isEmpty()) {
            event.getHook().sendMessage("You must mention a permission to adjust of the user you've mentioned.").setEphemeral(true).queue(invokeDeleteOnMessageResponse(deleteDelay));
            return null;
        }
        return mentionedMembersOptional.get();
    }

    @Override
    public String getName() {
        return "adjustuser";
    }

    @Override
    public String getDescription() {
        return "Use this command to adjust the mentioned user's permissions to use TobyBot commands for your server";
    }

    @Override
    public List<OptionData> getOptionData() {
        OptionData userOption = new OptionData(OptionType.STRING, USERS, "User(s) who you would like to adjust the permissions of.", true);
        OptionData permission = new OptionData(OptionType.STRING, PERMISSION_NAME, "What permission to adjust for the user", true);
        permission.addChoice(MUSIC.name(), MUSIC.name());
        permission.addChoice(MEME.name(), MEME.name());
        permission.addChoice(DIG.name(), DIG.name());
        permission.addChoice(SUPERUSER.name(), SUPERUSER.name());
        OptionData initiative = new OptionData(OptionType.INTEGER, MODIFIER, "modifier for the initiative command when used on your user");
        return List.of(userOption, permission, initiative);
    }
}

File: src/main/java/toby/command/commands/moderation/AdjustUserCommand.kt
package toby.command.commands.moderation

import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.entities.Mentions
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.interactions.commands.OptionMapping
import net.dv8tion.jda.api.interactions.commands.OptionType
import net.dv8tion.jda.api.interactions.commands.build.OptionData
import toby.command.CommandContext
import toby.command.ICommand.Companion.invokeDeleteOnMessageResponse
import toby.helpers.UserDtoHelper
import toby.jpa.dto.UserDto
import toby.jpa.service.IUserService
import java.util.*
import java.util.function.Consumer

class AdjustUserCommand(private val userService: IUserService) : IModerationCommand {
    private val PERMISSION_NAME = "name"
    private val USERS = "users"
    override fun handle(ctx: CommandContext?, requestingUserDto: UserDto, deleteDelay: Int?) {
        val event = ctx!!.event
        event.deferReply().queue()
        val member = ctx.member
        val guild = event.guild!!
        val mentionedMembers =
            channelAndArgumentValidation(event, requestingUserDto, guild.owner, member, deleteDelay!!)
                ?: return
        mentionedMembers.forEach(Consumer { targetMember: Member ->
            val targetUserDto = userService.getUserById(targetMember.idLong, targetMember.guild.idLong)
            //Check to see if the database contained an entry for the user we have made a request against
            if (targetUserDto != null) {
                val isSameGuild = requestingUserDto.guildId == targetUserDto.guildId
                val requesterCanAdjustPermissions =
                    UserDtoHelper.userAdjustmentValidation(requestingUserDto, targetUserDto) || member!!.isOwner
                if (requesterCanAdjustPermissions && isSameGuild) {
                    validateArgumentsAndUpdateUser(event, targetUserDto, member!!.isOwner, deleteDelay)
                    event.hook.sendMessageFormat("Updated user %s's permissions", targetMember.effectiveName)
                        .queue(invokeDeleteOnMessageResponse(deleteDelay))
                } else event.hook.sendMessageFormat(
                    "User '%s' is not allowed to adjust the permissions of user '%s'.",
                    member!!.effectiveName,
                    targetMember.effectiveName
                ).queue(invokeDeleteOnMessageResponse(deleteDelay))
            } else {
                createNewUser(event, targetMember, deleteDelay)
            }
        })
    }

    private fun validateArgumentsAndUpdateUser(
        event: SlashCommandInteractionEvent,
        targetUserDto: UserDto,
        isOwner: Boolean,
        deleteDelay: Int
    ) {
        val permissionOptional =
            Optional.ofNullable(event.getOption(PERMISSION_NAME)).map { obj: OptionMapping -> obj.asString }
        val modifierOptional = Optional.ofNullable(event.getOption(MODIFIER)).map { obj: OptionMapping -> obj.asInt }
        if (permissionOptional.isEmpty && modifierOptional.isEmpty) {
            event.hook.sendMessage("You did not mention a valid permission to update").setEphemeral(true)
                .queue(invokeDeleteOnMessageResponse(deleteDelay))
            return
        }
        modifierOptional.ifPresent { modifier: Int? -> targetUserDto.initiativeModifier = modifier!! }
        when (UserDto.Permissions.valueOf(permissionOptional.get().uppercase(Locale.getDefault()))) {
            UserDto.Permissions.MUSIC -> targetUserDto.musicPermission = !targetUserDto.musicPermission
            UserDto.Permissions.DIG -> targetUserDto.digPermission = !targetUserDto.digPermission
            UserDto.Permissions.MEME -> targetUserDto.memePermission = !targetUserDto.memePermission
            UserDto.Permissions.SUPERUSER -> {
                if (isOwner) targetUserDto.superUser = !targetUserDto.superUser
            }
        }
        userService.updateUser(targetUserDto)
    }

    private fun createNewUser(event: SlashCommandInteractionEvent, targetMember: Member, deleteDelay: Int) {
        //Database did not contain an entry for the user we have made a request against, so make one.
        val newDto = UserDto()
        newDto.discordId = targetMember.idLong
        newDto.guildId = targetMember.guild.idLong
        userService.createNewUser(newDto)
        event.hook.sendMessageFormat(
            "User %s's permissions did not exist in this server's database, they have now been created",
            targetMember.effectiveName
        ).queue(invokeDeleteOnMessageResponse(deleteDelay))
    }

    private fun channelAndArgumentValidation(
        event: SlashCommandInteractionEvent,
        requestingUserDto: UserDto?,
        member: Member?,
        guildOwner: Member?,
        deleteDelay: Int
    ): List<Member>? {
        if (!member!!.isOwner && !requestingUserDto!!.superUser) {
            event.hook.sendMessage(getErrorMessage(guildOwner!!.effectiveName)!!).setEphemeral(true)
                .queue(invokeDeleteOnMessageResponse(deleteDelay))
            return null
        }
        val mentionedMembersOptional =
            Optional.ofNullable(event.getOption(USERS)).map { obj: OptionMapping -> obj.mentions }
                .map { obj: Mentions -> obj.members }
        if (mentionedMembersOptional.isEmpty) {
            event.hook.sendMessage("You must mention 1 or more Users to adjust permissions of").setEphemeral(true)
                .queue(invokeDeleteOnMessageResponse(deleteDelay))
            return null
        }
        if (Optional.ofNullable(event.getOption(PERMISSION_NAME)).map { obj: OptionMapping -> obj.asString }.isEmpty) {
            event.hook.sendMessage("You must mention a permission to adjust of the user you've mentioned.")
                .setEphemeral(true).queue(invokeDeleteOnMessageResponse(deleteDelay))
            return null
        }
        return mentionedMembersOptional.get()
    }

    override val name: String
        get() = "adjustuser"
    override val description: String
        get() = "Use this command to adjust the mentioned user's permissions to use TobyBot commands for your server"
    override val optionData: List<OptionData>
        get() {
            val userOption =
                OptionData(OptionType.STRING, USERS, "User(s) who you would like to adjust the permissions of.", true)
            val permission =
                OptionData(OptionType.STRING, PERMISSION_NAME, "What permission to adjust for the user", true)
            permission.addChoice(UserDto.Permissions.MUSIC.name, UserDto.Permissions.MUSIC.name)
            permission.addChoice(UserDto.Permissions.MEME.name, UserDto.Permissions.MEME.name)
            permission.addChoice(UserDto.Permissions.DIG.name, UserDto.Permissions.DIG.name)
            permission.addChoice(UserDto.Permissions.SUPERUSER.name, UserDto.Permissions.SUPERUSER.name)
            val initiative =
                OptionData(OptionType.INTEGER, MODIFIER, "modifier for the initiative command when used on your user")
            return listOf(userOption, permission, initiative)
        }

    companion object {
        const val MODIFIER = "modifier"
    }
}
